### PR TITLE
Added Ring of Forging to Item Charges plugin.

### LINKED ISSUE #8865: Add "Ring of Forging" to the "Item Charges" plugin
**Is your feature request related to a problem? Please describe.**
Currently, Runelite's Item Charges plugin does not have any tracking features for the Ring of Forging. This item currently sees more use recently due to being introduced to Free-to-Play players.

**Describe the solution you'd like**
The Item Charges plugin can be updated to add useful "charge count" features similar to that of the "Dodgy Necklace" tracked features.

**Describe alternatives you've considered**
A notification (and/or audible warning) could be added for when Runelite sees the message "Your Ring of Forging has melted." in the chatbox.

### PR DESCRIPTION
**Pull request by zachtyll**
Fixes #8865 Add "Ring of Forging" to the "Item Charges" plugin.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 1]
```diff
@@ -395,6 +453,10 @@ private void updateJewelleryInfobox(ItemWithSlot type, Item[] items, EquipmentIn
 
 		if (itemWithCharge == null)
 		{
+			if (id == ItemID.RING_OF_FORGING && type == ItemWithSlot.RING_OF_FORGING)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Add this to the bottom with `else if`

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 2]
```diff
@@ -182,6 +197,24 @@ public void onChatMessage(ChatMessage event)
 			{
 				notifier.notify("Your Ring of Recoil has shattered");
 			}
+			else if (ringOfForgingCheckMatcher.find())
+			{
+				updateRingOfForgingCharges(Integer.parseInt(ringOfForgingCheckMatcher.group(1)));
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: this will error when the captured group is `one`. there should be a test for this too.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeOverlay.java [hunk 3]
```diff
@@ -69,6 +69,15 @@ public void renderItemOverlay(Graphics2D graphics, int itemId, WidgetItem itemWi
...
+		if (itemId == ItemID.RING_OF_FORGING)
+		{
+			if (!config.showRingOfForgingCount())
+			{
+				return;
+			}
+
+			charges = config.ringOfForging();
+		}
 		if (itemId == ItemID.DODGY_NECKLACE)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeOverlay.java]: `else if`

**Comment by Adam-** [comment]
Whats up with all of the basket/bellows changes to this?

**Comment by zachtyll** [comment]
> Whats up with all of the basket/bellows changes to this?

When I renumbered the configs the basket and the bellows appeared right in the middle of them. So rather than go back and change all the changes that had just been done they were put at the end positions.

**Comment by Adam-** [comment]
You should really not be resorting any of the config like this in this PR. What happened to that test you were writing? Ideally this should have tests for each matcher.

**Comment by zachtyll** [comment]
The new configurations could be appended to the end of the configurations file and the old ones put back into the order they are in the master repo. The tests are not located within the Itemcharges plugin folder where the git instance was set up so they did not get added to the commit. This can be changed. Would it be a good idea to include tests for the binding necklace at the same time, as the syntax for the tests would be almost identical?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java [hunk 4]
```diff
@@ -34,273 +34,313 @@
 public interface ItemChargeConfig extends Config
 {
 	@ConfigItem(
-		keyName = "veryLowWarningColor",
-		name = "Very Low Warning Color",
-		description = "Configure the color of the overlay when charges are very low",
-		position = 1
+			keyName = "veryLowWarningColor",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java]: This PR is applying the wrong tab depth to every single `@ConfigItem` annotation in this file. Please fix the indentation. (only one level further than the `@ConfigItem`)

tabs

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java [hunk 5]
```diff
@@ -34,273 +34,313 @@
...
-		description = ""
+			keyName = "explorerRing",
+			name = "",
+			description = ""
 	)
 	void explorerRing(int explorerRing);
 
+	@ConfigItem(
+			keyName = "showRingOfForgingCount",
+			name = "Ring of Forging Count",
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java]:
```suggestion
			name = "Show Ring of Forging Charges",
```

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java [hunk 6]
```diff
@@ -34,273 +34,313 @@
...
+		return true;
+	}
+
+	@ConfigItem(
+			keyName = "RingOfForgingNotification",
+			name = "Ring of Forging Notification",
+			description = "Configures if the Ring of Forging breaking notification is shown",
+			position = 21
+	)
+	default boolean ringOfForgingNotification()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargeConfig.java]: Move this notification default method below the `ringOfForging` hidden and setter methods.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 7]
```diff
@@ -75,10 +75,17 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern RING_OF_FORGING_CHECK_PATTERN = Pattern.compile(
+			"You can smelt ([0-9+]+|one) more pieces of iron ore before a ring melts\\.");
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Same indent issue here (and on the other added static vars), only increase the indentation level by one.
```suggestion
		"You can smelt ([0-9+]+|one) more pieces of iron ore before a ring melts\\.");
```

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 8]
```diff
@@ -225,6 +240,31 @@ else if (bindingNecklaceCheckMatcher.find())
 
 				updateBindingNecklaceCharges(charges);
 			}
+			else if (ringOfForgingCheckMatcher.find())
+			{
+
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Remove this empty line.
```suggestion
```

File: runelite-client/src/test/java/net/runelite/client/plugins/itemcharges/ItemChargePluginTest.java [hunk 9]
```diff
@@ -52,6 +52,16 @@
 	private static final String PROTECT_1 = "Your dodgy necklace protects you. <col=ff0000>It has 1 charge left.</col>";
 	private static final String BREAK = "Your dodgy necklace protects you. <col=ff0000>It then crumbles to dust.</col>";
 
+	private static final String CHECK_RING_OF_FORGING_FULL = "You can smelt 140 more pieces of iron ore before a ring melts.";
+	private static final String CHECK_RING_OF_FORGING_ONE = "You can smelt one more pieces of iron ore before a ring melts.";
+	private static final String USED_RING_OF_FORGING = "You retrieve a bar of iron.";
+	private static final String BREAK_RING_OF_FORGING = "Your Ring of Forging has melted.";
+
+	private static final String CHECK_BINDING_NECKLACE_FULL = "You have 16 charges left before your Binding necklace disintegrates.";
```
Comment (Nightfirecat) [inline runelite-client/src/test/java/net/runelite/client/plugins/itemcharges/ItemChargePluginTest.java]: Binding necklace tests should be added in a separate commit, if not a fully separate PR.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 10]
```diff
@@ -75,10 +75,17 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern RING_OF_FORGING_CHECK_PATTERN = Pattern.compile(
+		"You can smelt ([0-9+]+|one) more pieces of iron ore before a ring melts\\.");
+	private static final Pattern RING_OF_FORGING_USED_PATTERN = Pattern.compile(
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: This is a static string with no variance, so use a `String` and use `String#contains()` to check for it within the chat message subscriber. (see `BINDING_BREAK_TEXT`'s usage for an example)

```java
private static final String RING_OF_FORGING_USED_TEXT = "You retrieve a bar of iron.";
```

Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Assuming the test is correct, `equals` should be fine.

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: The same change can be made to `RING_OF_FORGING_USED_PATTERN`, it should be a String instead.

**Comment by Adam-** [comment]
Your test has a check for: `"You can smelt one more pieces of iron ore before a ring melts."` - did you get this from the game? It is not grammatically correct

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 11]
```diff
@@ -75,10 +75,15 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern RING_OF_FORGING_CHECK_PATTERN = Pattern.compile(
+		"You can smelt ([0-9+]+|one) more pieces of iron ore before a ring melts\\.");
+	private static final String RING_OF_FORGING_USED_TEXT = "You retrieve a bar of iron.";
+	private static final String RING_OF_FORGING_BREAK_TEXT = "Your Ring of Forging has melted.";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: This message has some color applied, so this is not the correct exact string.

Ring of forging melt text

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 12]
```diff
@@ -75,10 +75,15 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern RING_OF_FORGING_CHECK_PATTERN = Pattern.compile(
+		"You can smelt ([0-9+]+|one) more pieces of iron ore before a ring melts\\.");
```
Comment (Alexsuperfly) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: when only one charge remains pieces -> piece

**Comment by zachtyll** [comment]
> Your test has a check for: "You can smelt one more pieces of iron ore before a ring melts." - did you get this from the game? It is not grammatically correct

I just re-checked and it says "You can smelt one more piece of iron ore Before a ring melts.". Therefore it won't work as intended. When I tested this in game it was however not noticeable since we never see the item charge at 0 as the ring melts away. I could change the pattern to decrease to 0 at the last charge.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java (hunk 11 above)
Comment (zachtyll) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Oh alright! I wasn't aware that colour was coded into the string itself. That explains why the recoil text looks odd too. Do you recommend using the same syntax and if so, which value is the purple text?

Comment (Nightfirecat) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: I'd recommend using the same syntax as the recoil text. Sadly, I don't know the value (not my screenshot), but the color from the image is `rgb(127, 0, 127)`, which would be `"<col=7f007f>"`.

Comment (zachtyll) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Ok thanks! I just checked in game and the next Ring of Forging has 140 charges so the pattern matcher must have caught the message. I will alter the code anyway for the sake of correctness.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java (hunk 12 above)
Comment (zachtyll) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Thank you! I added the greedy quantifier '?' we see in the other pattern matchers when the syntax changes on the last charge.

Comment (Adam-) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: You should be adding test cases which contain the actual text the game uses. At least one of your current test cases is just made up.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 13]
```diff
@@ -75,10 +75,15 @@
 	private static final Pattern BINDING_USED_PATTERN = Pattern.compile(
 		"You bind the temple's power into (mud|lava|steam|dust|smoke|mist) runes\\.");
 	private static final String BINDING_BREAK_TEXT = "Your Binding necklace has disintegrated.";
+	private static final Pattern RING_OF_FORGING_CHECK_PATTERN = Pattern.compile(
+		"You can smelt ([0-9+]+|one) more pieces? of iron ore before a ring melts\\.");
+	private static final String RING_OF_FORGING_USED_TEXT = "You retrieve a bar of iron.";
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Isn't this message the same for smelting a bar of iron whether or not you have a ring of forging equipped? As-is, it would change the charge count even if you have no ring on.

Comment (zachtyll) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Yes you are right! Is there a way to detect the itemID for what is equipped in the ring slot?

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java [hunk 14]
```diff
@@ -32,14 +32,8 @@
...
-import net.runelite.api.ChatMessageType;
-import net.runelite.api.Client;
-import net.runelite.api.EquipmentInventorySlot;
-import net.runelite.api.InventoryID;
-import net.runelite.api.Item;
-import net.runelite.api.ItemContainer;
-import net.runelite.api.ItemID;
-import net.runelite.api.Varbits;
+
+import net.runelite.api.*;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargePlugin.java]: Don't use wildcard imports: https://github.com/runelite/runelite/wiki/Code-Conventions#imports

**Comment by zachtyll** [comment]
The slayer plugin makes use of getters and setters for testing the different bracelets one would use when slaying. Is this a recommended way of implementing tests for things that need to be worn or is there a better way to do it?
